### PR TITLE
[HmGui] Layout system

### LINKED ISSUE #166: [HmGui] Implement layouting
# HmGui layout design.

The main building block of the Gui system is a widget. It contains several parameters used in styling and layout. Each widget can be one of 4 types:
1. Image. To visualize image.
2. Rect. To draw rectangle.
3. Text. To draw a text with specified font and color.
4. Container. Groups child widgets inside parent widget. Defines layout type for positioning and sizing children.

Widgets of the container type are used to organize tree hierarchy of Gui blocks. Layout algorithm uses this hierarchy to position and size widgets.

## Widget

Widget has following parameters:
1. Width: **Option\<f32>/None**, height: **Option\<f32>/None**. User defined optional values to specify widget's fixed width and/or height. If fixed size set for the container widget then layout algorithm will attempt to fit all its children inside this size. If that's not possible then only part of the children will be visible.
2. Docking: **u8/0**. 4 bits are used for top, bottom, left and right widget docking in a parent container. If docking is not specified then widget will be centered inside parent with minimal possible size.
3. Margin: **Vec4/(0,0,0,0)**. Free space around widget.
4. Background color: **Vec4/(0,0,0,0)**. RGBA color of the widget's background.
5. Border width: **f32/0**. Border lays around widget's sides and together with margin defines space around the widget.
6. Border color: **Vec4/(0,0,0,0)**. RGBA color of the widget's border.

## Container

Container widget has following parameters:
1. Children: **Vec\<Widget>/Empty**. Child widgets.
2. Layout type: **enum(Stack,Vertical,Horizontal)**. Specifies how child widgets are organized inside the container.
    - Stack: widgets are organized in a stack laying on each other
    - Vertical: widgets are organized in a column
    - Horizontal: widgets are organized in a row
3. Docking: **u8/0**. 4 bits are used for top, bottom, left and right children docking. If docking is not specified then children will be centered inside this container with minimal possible size. For vertical and horizontal layout type children can override horizontal and vertical docking of the container respectively. For the stack layout type children can override both docking directions. Override happens only if child widget has docking equal to 0.
4. Padding: **Vec4/(0,0,0,0)**. Internal free space between container widget sides and children.
5. Spacing: **f32/0**. Additional free space between the children. Only for the Vertical and Horizontal layout types.

## Text

Text widget has following fields:
1. Text: **String/Empty**. Text to visualize.
2. Font name: **String/Some default font**. Freetype font.
3. Color: **Vec4/(0,0,0,1)**. Text font.
4. Expand: **bool/true**. Specifies if text should automatically scale when widget change size.

### PR DESCRIPTION
**Pull request by HaronK**
Closes #166

### CODE REVIEW

File: engine/lib/phx/src/ui/hmgui/README.md
Comment (Flatfingers) [inline engine/lib/phx/src/ui/hmgui/README.md:1]: # HmGui

HmGui (Hybrid Mode GUI) is a Graphical User Interface (GUI) framework that combines the best parts of the Immediate Mode and Retained Mode GUI models while minimizing their downsides.

The basic element of HmGui is the **widget**. There are three fundamental types of widgets: Text, Rect (rectangle), and Container. All more complex HmGui elements, such as Label, Textbox, Button, Checkbox, and Menu, are composed of combinations of these three basic widget types.

A widget represents an object covering a rectangular area to which can be assigned several different types of layout (positioning), styling (appearance), and window scrolling properties.

Images will be supported as a special property of the widget.

These features of HmGui are sufficient to accurately build complex GUI structures.

## Layout System

The HmGui layout system is based on the following property assignments:

1. Layout model of the container: Stack, Horizontal, Vertical
2. Alignment of child elements.
   - horizontal: Center, Left (default), Right, Stretch
   - vertical: Center, Top (default), Right, Stretch
3. Size specification: Fixed, Percent
4. Decorations: padding, spacing, margin and border

All of these properties can affect how elements are placed on the screen relative to each other and to the parent.

### Layout Model

A Container is a widget that may or may not have child elements attached to it. If it does, those child elements will be drawn such that they are positioned inside their parent Container according to the layout model assigned to that Container.

HmGui supports three layout models for arranging child elements in a Container:

- **Stack** (elements are drawn on top of each other and may overlap)
- **Horizontal** (elements are drawn horizontally from left to right without overlapping)
- **Vertical** (elements are drawn vertically from top to bottom without overlapping)

In addition to a parent Container's layout model, the positioning of elements in a Container is also affected by several properties of child elements:

- **alignment** - telling an element to attach one or more of its four sides (Left, Right, Top and Bottom) to its parent's side, to be centered in its parent, or to stretch as far as possible (inside its parent) either vertically or horizontally
- **fixed/percent size** - telling an element to either:
  - span a fixed size in pixels, or
  - expand, if possible, to a percentage of its parent's size
- **padding** - transparent space in pixels between the Container's enclosing Rect and its children
- **spacing** - transparent space in pixels between each child in either a Horizontal or Vertical layout
- **border** - styled space in pixels outside a Container rect
- **margin** - transparent space in pixels outside a Container and its border if one is specified

All elements may have alignment and fixed/percent size property values assigned to them.

Containers may also have decoration property values assigned to them, which may affect how child elements are laid out inside them.

Example:
```lua
Gui:beginHorizontalContainer()
Gui:setFixedSize(100, 100)

Gui:rect()
Gui:setFixedHeight(10)
Gui:setFixedWidth(20)

Gui:rect()
Gui:setFixedHeight(10)
Gui:setFixedWidth(20)

Gui:rect()
Gui:setFixedHeight(10)
Gui:setFixedWidth(20)

Gui:endContainer()
```
All three Rect elements will be drawn in a horizontal row, in order from left to right next to each other, with the entire group of three Rects centered in their parent Container. Each element will retain its fixed height and width.

### Alignment

Alignment is a property that can be used to position an element in relation to any of the four sides of its parent Container, or to the element beside the selected child element in the designated direction.

There are two dimensions along which elements can be aligned: Horizontal (width) and Vertical (height). In both alignment dimensions, one of six parameters can be specified: Center, Left (default), Right, Top (default), Bottom, and Stretch.

The HmGui programmer can set alignment for both dimensions independently.

Stretching an element in the horizontal or vertical dimension by alignment always has priority over fixed/percent size in that dimension. If both alignment and a fixed or percent size are specified for an element, the alignment effect will be applied.

Example:
```lua
Gui:beginStackContainer()
Gui:setFixedSize(100, 100)

Gui:rect()
Gui:setFixedWidth(20)
Gui:setVerticalAlignment(AlignVertical.Stretch)

Gui:rect()
Gui:setFixedHeight(20)
Gui:setHorizontalAlignment(AlignHorizontal.Stretch)

Gui:rect()
Gui:setAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)

Gui:endContainer()
```
Because this Container has been told to use the Stack layout model (through the command `Gui:beginStackContainer()`, these three Rects will overlap. There will be one tall, skinny Rect on the left; one short, wide Rect at the top; and one Rect that fills the Container.

Additionally, the commands `setChildrenHorizontalAlignment()` and `setChildrenVerticalAlignment()` may be applied to Containers. These will impose one alignment effect on all of that Container's child elements. For example, to align all children to the left inside a container using the Vertical layout model, users can do the following:

```lua
Gui:beginVerticalContainer()
Gui:setFixedSize(100, 100)
Gui:setChildrenHorizontalAlignment(AlignHorizontal.Left)

Gui:rect()
Gui:setFixedWidth(20)

Gui:rect()
Gui:setFixedWidth(40)

Gui:rect()
Gui:setFixedWidth(30)

Gui:endContainer()
```

All of the three child Rect elements in this example will be drawn stacked on top of each other, in order from top to bottom, with each element's left side touching the left side of their parent container, and the group of three elements will be centered vertically by default (since no explicit Vertical alignment was specified).

Note: `setChildrenHorizontalAlignment()` will accept only the parameters AlignHorizontal.Center, AlignHorizontal.Left, AlignHorizontal.Right, and AlignHorizontal.Stretch; and `setChildrenVerticalAlignment()` will accept only the parameters AlignVertical.Center, AlignVertical.Top, AlignVertical.Bottom, and AlignVertical.Stretch.

#### Advanced Alignment

`Gui:setChildren*Alignment()` methods can be used for simple alignment of the Container's children. This alignment will be applied only for the children without explicit child alignment specification and explicit fixed/percent size (children alignment has lower priority than child fixed/percent size in contrast to child alignment that has higher priority).

Example:
```lua
Gui:beginHorizontalContainer()
Gui:setFixedSize(100, 100)
Gui:setChildrenHorizontalAlignment(AlignHorizontal.Stretch)

Gui:rect()

Gui:rect()
Gui:setFixedWidth(20)

Gui:rect()

Gui:endContainer()
```
In this example middle Rect will keep width of 20 pixels while first and last Rects will be stretched proportionally in the horizontal dimension to fill remaining parent's width.

### Fixed/Percent Size

Fixed size specifies a length in pixels.

It is possible for a child element to be defined to have a greater fixed size (height and/or width) than its parent Container. The part of the element outside its parent Container's extents will be clipped unless the Container has the scrollable property in that dimension set to true.

Percent size specifies length in terms of a percentage of the parent container's size.

In the case of a Container with a Horizontal or Vertical layout model specified, if a percent size is applied in the specified dimension, then that percentage will be calculated against the total parent size minus the size of all fixed sized elements and minus any padding in that dimension.

Example:
```lua
Gui:beginHorizontalContainer()
Gui:setFixedWidth(100)

Gui:rect()
Gui:setFixedWidth(20)

Gui:rect()
Gui:setPercentWidth(75)

Gui:rect()
Gui:setPercentWidth(25)

Gui:endContainer()
```
In this case the width of the second rect will be 60 and third one - 20 (instead of 75 and 25).

Any element size defined as a percent size may be reduced, possibly to 0, to prevent the total size of all elements in that dimension from exceeding their parent Container's size.

### Decorations

HmGui elements can be optionally assigned an outer margin (free space outside itself) and a border. These decorations are drawn outside of the element's enclosing rectangle. If the element is a Container, when fixed or percent sizes are specified for child elements, they are applied only to the current interior size of their parent Container.

Other decorations that Containers can have are inner padding (free space) inside its enclosing rectangle, and (for Containers with Horizontal or Vertical layout) spacing in between each child element in the dimension specified by that layout model.

Example:
```lua
Gui:beginHorizontalContainer()
Gui:setPadding(10)
Gui:setSpacing(5)

Gui:rect()
Gui:setFixedWidth(20)
Gui:setMargin(5, 5)

Gui:rect()
Gui:setPercentWidth(100)
Gui:setBorderWidth(2)

Gui:rect()
Gui:setFixedWidth(25)

Gui:endContainer()
```

## Styling

In addition to properties affecting layout, some elements may be assigned additional properties affecting their appearance. These are referred to as styling properties.

The styling properties to be implemented initially are:

- border-color
- background-color
- background-image

Color is specified as an ordered set of four floating-point values, with each value constrained to the range [0.0 : 1.0]. The four values of a Color are: Red, Green, Blue, Alpha (opacity).

Image is specified as an image format data blob whose contents have been read from a named file organized in one of the following graphics formats: JPG/JPEG, GIF, PNG.

Comment (HaronK) [reply engine/lib/phx/src/ui/hmgui/README.md:1]: Updated

**Comment by Flatfingers** [comment]
local Test = require('States.Application')

local rng = RNG.FromTime()

local useRenderer = true
local drawExitMenu = false

local station_options = 1

function Test:onInit()
    self.renderer = Renderer()
end

function Test:onInput() end

--** Begin game menu functions **--

function Test:showMarketOverlay()
    Gui:beginVerticalContainer()
    Gui:setAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)
    Gui:setChildrenAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)
    self:showMarketOverlayTitle()
    self:showMarketOverlayMenuBar()
    self:showMarketOverlayDivider()
    self:showMarketOverlayData()
    Gui:endContainer()
end

function Test:showMarketOverlayTitle()
    Gui:beginStackContainer()
    Gui:setPercentSize(100, 3)
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Stretch)
    if station_options == 1 then
        Gui:textEx(Cache.Font('Iceland', 64), 'MARKETPLACE', 1.0, 1.0, 1.0, 1.0)
    elseif station_options == 2 then
        Gui:textEx(Cache.Font('Iceland', 64), 'SHIP CONFIGURATION', 1.0, 1.0, 1.0, 1.0)
    elseif station_options == 3 then
        Gui:textEx(Cache.Font('Iceland', 64), 'OCCUPATIONAL', 1.0, 1.0, 1.0, 1.0)
    else
        Gui:textEx(Cache.Font('Iceland', 64), 'WTF?', 1.0, 1.0, 1.0, 1.0)
    end
    Gui:endContainer()
end

function Test:showMarketOverlayMenuBar()
    Gui:beginHorizontalContainer() -- divide horizontal menu bar area into three sections: location, menu buttons, and money
    Gui:setPercentSize(100, 6)
    Gui:setAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)
    Gui:beginVerticalContainer() -- section 1 is the player ship's current location
    Gui:setPercentSize(15, 100)
    Gui:setAlignment(AlignHorizontal.Left, AlignVertical.Center)
    Gui:setChildrenAlignment(AlignHorizontal.Center, AlignVertical.Center)
    Gui:textEx(Cache.Font('Iceland', 32), 'Docked at', 1.0, 1.0, 1.0, 1.0)
    Gui:textEx(Cache.Font('Iceland', 32), 'Titan Station', 1.0, 1.0, 1.0, 1.0)
    Gui:textEx(Cache.Font('Iceland', 24), 'Parnell System', 1.0, 1.0, 1.0, 1.0)
    Gui:endContainer()
    Gui:beginStackContainer() -- section 2 contains the menu buttons (such as "Commodities") associated with this market function (such as "Market")
    Gui:setPercentSize(70, 100)
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Center)
    Gui:textEx(Cache.Font('Iceland', 24), '(menu buttons)', 1.0, 1.0, 1.0, 1.0)
    Gui:endContainer()
    Gui:beginStackContainer() -- section 3 contains the player's current financial balance
    Gui:setPercentSize(15, 100)
    Gui:setAlignment(AlignHorizontal.Right, AlignVertical.Center)
    Gui:textEx(Cache.Font('Iceland', 24), '$123,456,789', 1.0, 1.0, 1.0, 1.0)
    Gui:endContainer()
    Gui:endContainer()
end

function Test:showMarketOverlayDivider()
    Gui:beginStackContainer()
    Gui:setPercentSize(100, 1)
    Gui:setAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)
    -- It would be nice to have a divider() method that displays a styled horizontal rule, but a rect can work.
    Gui:rect(1.0, 1.0, 1.0, 1.0)
    Gui:setFixedHeight(2)
    Gui:setHorizontalAlignment(AlignHorizontal.Stretch)
    Gui:endContainer()
end

function Test:showMarketOverlayData()
    Gui:beginStackContainer()
    Gui:setPercentSize(100, 90)
    Gui:setAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)
    -- This is where the bulk of the UI code will go to draw the many various Marketplace screens
    Gui:endContainer()
end

--** Begin game control menu functions **--

function Test:showMetrics()
    Gui:beginStackContainer()
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Bottom)
    Gui:setPadding(6, 6)
    Gui:textEx(Cache.Font('Exo2', 12), format("FPS: %.2f", 1.0 / self.dt), 1.0, 1.0, 1.0, 1.0)
    Gui:endContainer()
end

function Test:showCtrlDialogInner()
    Gui:beginHorizontalContainer()
    Gui:setBorderWidth(2)
    Gui:pushTextColor(1.0, 1.0, 1.0, 1.0)
    Gui:pushFont(Cache.Font('Exo2Bold', 18))
    if Gui:button("Cancel") then
        drawExitMenu = false
    end
    Gui:setSpacing(24)
    if Gui:button("Quit") then
        Test:quit()
    end
    Gui:endContainer()
end

function Test:showCtrlDialog()
    Gui:beginStackContainer()
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Center)
    Gui:beginWindow("Game Control", InputInstance)
    Gui:setBorderWidth(2)
    Gui:textEx(Cache.Font('Iceland', 20), 'Game Control', 0.6, 0.7, 1.0, 1.0)
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Center)
    Gui:setSpacing(24)
    self:showCtrlDialogInner()
    Gui:endWindow()
    Gui:endContainer()
end

function Test:onUpdate(dt)
    Profiler.Begin('Gui:update')
    if InputInstance:isPressed(Button.KeyboardEscape) then
        drawExitMenu = true
    end
    Gui:beginGui(self.resX, self.resY, InputInstance)
    if drawExitMenu then
        self:showCtrlDialog()
    else
        self:showMarketOverlay()
        self:showMetrics()
    end
    Gui:endGui(InputInstance)
    Profiler.End()
end

function Test:onDraw()
    if useRenderer then
        self.renderer:start(self.resX, self.resY)
        Viewport.Push(0, 0, self.resX, self.resY, true)
        Gui:draw()
        Viewport.Pop()
        self.renderer:stop()
        self.renderer:present(0, 0, self.resX, self.resY)
    else
        Gui:draw()
    end
end

return Test

**Comment by Flatfingers** [comment]
```
local Test = require('States.Application')

local rng = RNG.FromTime()

local useRenderer = true
local drawExitMenu = false

local station_options = 1

local player_account = 123456789

function Test:onInit()
    self.renderer = Renderer()
end

function Test:onInput() end

--** Begin game menu functions **--

function Test:showMarketOverlay()
    Gui:beginVerticalContainer()
    Gui:setBorder(2, 0.0, 1.0, 0.0, 1.0)
    Gui:setBgColor(0.3, 0.3, 0.3, 0.8)
    Gui:setAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)
    Gui:setChildrenHorizontalAlignment(AlignHorizontal.Stretch)

    self:showMarketOverlayTitle()
    self:showMarketOverlayMenuBar()
    Gui:horizontalDivider(2, 1.0, 1.0, 1.0, 1.0)
    self:showMarketOverlayData()

    Gui:endContainer()
end

function Test:showMarketOverlayTitle()
    if station_options == 1 then
        Gui:textEx(Cache.Font('Iceland', 64), 'MARKETPLACE', 1.0, 1.0, 1.0, 1.0)
    elseif station_options == 2 then
        Gui:textEx(Cache.Font('Iceland', 64), 'SHIP CONFIGURATION', 1.0, 1.0, 1.0, 1.0)
    elseif station_options == 3 then
        Gui:textEx(Cache.Font('Iceland', 64), 'OCCUPATIONAL', 1.0, 1.0, 1.0, 1.0)
    else
        Gui:textEx(Cache.Font('Iceland', 64), 'WTF?', 1.0, 1.0, 1.0, 1.0)
    end
    Gui:setHorizontalAlignment(AlignHorizontal.Center)
    Gui:setMargin(5, 5)
end

function Test:showMarketOverlayMenuBar()
    -- Divide the horizontal menu bar area into three sections: location, menu buttons, and money
    Gui:beginHorizontalContainer()
--    Gui:setHorizontalAlignment(AlignHorizontal.Stretch) -- this doesn't actually seem to do anything whether it's on or off
    Gui:setChildrenVerticalAlignment(AlignVertical.Stretch)
    Gui:setChildrenHorizontalAlignment(AlignHorizontal.Center)

    -- Section 1 is the player ship's current location
    Gui:beginVerticalContainer()
--    Gui:setHorizontalAlignment(AlignHorizontal.Left)
--    Gui:setPercentWidth(20)
--    Gui:setFixedWidth(250)
    Gui:setBorder(2, 1.0, 0.0, 0.0, 1.0)
    Gui:setPadding(15, 15)
    Gui:setChildrenAlignment(AlignHorizontal.Center, AlignVertical.Center)

    Gui:textEx(Cache.Font('Iceland', 32), 'Docked at', 1.0, 1.0, 1.0, 1.0)
    Gui:textEx(Cache.Font('Iceland', 32), 'Titan Station', 1.0, 1.0, 1.0, 1.0)
    Gui:textEx(Cache.Font('Iceland', 24), 'Parnell System', 1.0, 1.0, 1.0, 1.0)

    Gui:endContainer()

    -- Section 2 contains the menu buttons (such as "Commodities") associated with this market function (such as "Market")
    Gui:beginHorizontalContainer()
--    Gui:setHorizontalAlignment(AlignHorizontal.Stretch)
--    Gui:setPercentWidth(70)
--    Gui:setFixedWidth(1390)
    Gui:setBorder(2, 1.0, 0.0, 0.0, 1.0)
    Gui:setPadding(15, 15)
    Gui:setChildrenAlignment(AlignHorizontal.Center, AlignVertical.Center)

    Gui:pushTextColor(1.0, 0.0, 1.0, 1.0)
    Gui:pushFont(Cache.Font('Exo2Bold', 28))

    if Gui:button("Commodities") then
        player_account = 123456789
    end
    Gui:setFixedSize(280, 60)
    Gui:setBorder(2, 1.0, 0.0, 1.0, 1.0)
    if Gui:button("Data / Intel") then
        player_account = 123456
    end
    Gui:setFixedSize(280, 60)
    Gui:setBorder(2, 1.0, 0.0, 1.0, 1.0)
    if Gui:button("Exchange Orders") then
        player_account = 123
    end
    Gui:setFixedSize(280, 60)
    Gui:setBorder(2, 1.0, 0.0, 1.0, 1.0)
    if Gui:button("Blank") then
        player_account = 1
    end
    Gui:setFixedSize(280, 60)
    Gui:setBorder(2, 1.0, 0.0, 1.0, 1.0)

    Gui:popStyle(2)

    Gui:endContainer()

    -- Section 3 contains the player's current financial balance
    Gui:beginStackContainer()
--    Gui:setHorizontalAlignment(AlignHorizontal.Right)
--    Gui:setPercentWidth(15)
--    Gui:setFixedWidth(250)
    Gui:setBorder(2, 1.0, 0.0, 0.0, 1.0)
    Gui:setPadding(15, 15)
    Gui:setChildrenAlignment(AlignHorizontal.Center, AlignVertical.Center)

    Gui:textEx(Cache.Font('Iceland', 24), "$" .. player_account, 1.0, 1.0, 1.0, 1.0)

    Gui:endContainer()

    Gui:endContainer() -- end horizontal menu bar area
end

function Test:showMarketOverlayData()
    -- This is where the bulk of the UI code will go to draw the many various Marketplace elements
    Gui:beginHorizontalContainer()
    Gui:setBorder(2, 1.0, 1.0, 0.0, 1.0)
    Gui:setBgColor(0.1, 0.1, 0.1, 0.8)
    Gui:setAlignment(AlignHorizontal.Stretch, AlignVertical.Stretch)

    Gui:beginVerticalContainer() -- market category buttons in left-hand panel
    Gui:setBorder(2, 1.0, 0.0, 0.0, 1.0)
    Gui:setPercentWidth(15)
    Gui:setVerticalAlignment(AlignVertical.Stretch)



    Gui:endContainer() -- end market category buttons panel

    Gui:beginVerticalContainer()
    Gui:setBorder(2, 0.0, 1.0, 0.0, 1.0)
    Gui:setPercentWidth(85)
    Gui:setVerticalAlignment(AlignVertical.Stretch)



    Gui:endContainer()

--    Gui:endContainer() -- end market data area
end

--** Begin game control menu functions **--

function Test:showMetrics()
    Gui:beginStackContainer()
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Bottom)
    Gui:setPadding(6, 6)
    Gui:textEx(Cache.Font('Exo2', 12), format("FPS: %.2f", 1.0 / self.dt), 1.0, 1.0, 1.0, 1.0)
    Gui:endContainer()
end

function Test:showCtrlDialogInner()
    Gui:beginHorizontalContainer()
    Gui:setBorderWidth(2)
    Gui:pushTextColor(1.0, 1.0, 1.0, 1.0)
    Gui:pushFont(Cache.Font('Exo2Bold', 18))
    if Gui:button("Cancel") then
        drawExitMenu = false
    end
    Gui:setSpacing(24)
    if Gui:button("Quit") then
        Test:quit()
    end
    Gui:popStyle(2)
    Gui:endContainer()
end

function Test:showCtrlDialog()
    Gui:beginStackContainer()
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Center)
    Gui:beginWindow("Game Control", InputInstance)
    Gui:setBorderWidth(2)
    Gui:textEx(Cache.Font('Iceland', 20), 'Game Control', 0.6, 0.7, 1.0, 1.0)
    Gui:setAlignment(AlignHorizontal.Center, AlignVertical.Center)
    Gui:setSpacing(24)
    self:showCtrlDialogInner()
    Gui:endWindow()
    Gui:endContainer()
end

function Test:onUpdate(dt)
    Profiler.Begin('Gui:update')
    if InputInstance:isPressed(Button.KeyboardEscape) then
        drawExitMenu = true
    end
    Gui:beginGui(self.resX, self.resY, InputInstance)
    if drawExitMenu then
        self:showCtrlDialog()
    else
        self:showMarketOverlay()
        self:showMetrics()
    end
    Gui:endGui(InputInstance)
    Profiler.End()
end

function Test:onDraw()
    if useRenderer then
        self.renderer:start(self.resX, self.resY)
        Viewport.Push(0, 0, self.resX, self.resY, true)
        Gui:draw()
        Viewport.Pop()
        self.renderer:stop()
        self.renderer:present(0, 0, self.resX, self.resY)
    else
        Gui:draw()
    end
end

return Test
```

File: engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs
Comment (dgavedissian) [inline engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs:8]: Can we pre-emptively add DVec3/DVec4 and IVec3 here too?

File: engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs
Comment (dgavedissian) [inline engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs:11]: Do you mind adding the remaining vector types as well as Box3? I did it in my Rapier PR, but I think it would be good for you to add it here, as you're merging this soon: https://github.com/Limit-Theory-Redux/ltheory/pull/92/files#diff-3182aa17288a8515eed368a0d87ab758a0b02ab3a4f9c6f55b743351a84a13d1

File: engine/lib/phx/src/ui/hmgui/container.rs
Comment (dgavedissian) [inline engine/lib/phx/src/ui/hmgui/container.rs:33]: Can we answer this todo before merging, or is it something for later? If it's for later, lets track it somewhere as part of an issue.

**Review by dgavedissian — APPROVED** [review]
I think we should wait until @Flatfingers approves before this is merged, but I'm happy with the Rust code and the new Lua interface looks neat too. Got a few minor comments.

File: engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs
Comment (HaronK) [reply engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs:13]: Done

File: engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs
Comment (HaronK) [reply engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs:17]: Done except changes in `COPY_TYPES` array - only ***Vec2** types can be copuable

File: engine/lib/phx/src/ui/hmgui/container.rs
Comment (HaronK) [reply engine/lib/phx/src/ui/hmgui/container.rs]: Not used at the moment - removed

**Review by Flatfingers — APPROVED** [review]

**Pull request merged**
